src/command/subgraph/introspect.rs

```diff
@@ -2,48 +2,94 @@ use reqwest::blocking::Client;
 use saucer::{clap, Parser};
 use serde::Serialize;
 use std::collections::HashMap;
-use url::Url;
+
+use crate::{command::output::JsonOutput, options::IntrospectOpts};
 
 use rover_client::{
     blocking::GraphQLClient,
-    operations::subgraph::introspect::{self, SubgraphIntrospectInput},
+    operations::subgraph::introspect::{self, SubgraphIntrospectInput, SubgraphIntrospectResponse},
 };
 
 use crate::command::RoverOutput;
-use crate::utils::parsers::parse_header;
 use crate::Result;
 
 #[derive(Debug, Serialize, Parser)]
 pub struct Introspect {
-    /// The endpoint of the subgraph to introspect
-    #[serde(skip_serializing)]
-    pub endpoint: Url,
-
-    /// headers to pass to the endpoint. Values must be key:value pairs.
-    /// If a value has a space in it, use quotes around the pair,
-    /// ex. -H "Auth:some key"
-
-    // The `name` here is for the help text and error messages, to print like
-    // --header <key:value> rather than the plural field name --header <headers>
-    #[clap(name="key:value", multiple=true, long="header", short='H', parse(try_from_str = parse_header))]
-    #[serde(skip_serializing)]
-    pub headers: Option<Vec<(String, String)>>,
+    #[clap(flatten)]
+    opts: IntrospectOpts,
 }
 
 impl Introspect {
-    pub fn run(&self, client: Client) -> Result<RoverOutput> {
-        let client = GraphQLClient::new(&self.endpoint.to_string(), client)?;
+    pub fn run(&self, client: Client, json: bool) -> Result<RoverOutput> {
+        if self.opts.watch {
+            self.exec_and_watch(&client, json)?;
+            Ok(RoverOutput::EmptySuccess)
+        } else {
+            let response = self.exec(&client, true)?;
+            Ok(RoverOutput::Introspection(response.result))
+        }
+    }
+
+    pub fn exec(&self, client: &Client, should_retry: bool) -> Result<SubgraphIntrospectResponse> {
+        let client = GraphQLClient::new(&self.opts.endpoint.to_string(), client.clone())?;
 
         // add the flag headers to a hashmap to pass along to rover-client
         let mut headers = HashMap::new();
-        if let Some(arg_headers) = &self.headers {
+        if let Some(arg_headers) = &self.opts.headers {
             for (header_key, header_value) in arg_headers {
                 headers.insert(header_key.to_string(), header_value.to_string());
             }
         };
 
-        let introspection_response = introspect::run(SubgraphIntrospectInput { headers }, &client)?;
+        Ok(introspect::run(
+            SubgraphIntrospectInput { headers },
+            &client,
+            should_retry,
+        )?)
+    }
+
+    pub fn exec_and_watch(&self, client: &Client, json: bool) -> Result<RoverOutput> {
```

Comment:
I _think_ this and the `exec_and_watch` above are the same; worth putting into a util fn shared between the two? Or, do you expect subgraphs and supergraphs will differ in how they use `exec_and_watch`?

Reply:
yeah this could be separated out

Reply:
done!